### PR TITLE
strip whitespace when it uses tabs

### LINKED ISSUE #61: strip whitespace when it uses tabs, too
#57 removed leading whitespace (see #51), but it only handles spaces, not tabs. See this comment:

This is probably good enough, but it only strips leading *spaces*.

I feel like 

```
\t\tFoo
\t\tBar
```

should also work. To make it work, we would want to do

```rust
take_while(|c| c.is_whitespace())
```

and then collect into a *String*, I think. Not as simple, though.

_Originally posted by @nikomatsakis in https://github.com/dada-lang/dada/pull/57#discussion_r785300730_

### PR DESCRIPTION
**Pull request by XFFXFF**
This pr attemps to fix #61 . 
I don't know if there is a better way to do this, feel free to give suggestions.

### CODE REVIEW

**Comment by XFFXFF** [comment]
I didn't add a test like
```
\t\tFoo
\t\tBar
```
Because "\t" will be lexed as "\\\t" now, which is not a whitespace.

File: components/dada-validate/src/validate/validator.rs [hunk 1]
```diff
@@ -540,22 +541,49 @@ impl<'me> Validator<'me> {
...
-        .lines()
-        .filter(|&line| !line.trim().is_empty())
-        .collect::<Vec<&str>>();
-
-    let indent = non_empty_lines
-        .iter()
-        .map(|line| line.chars().into_iter().take_while(|c| *c == ' ').count())
+
+    // Split string into lines and filter out empty lines.
+    let non_empty_line_iter = s.lines().filter(|&line| !line.trim().is_empty());
```
Comment (nikomatsakis) [inline components/dada-validate/src/validate/validator.rs:552]: A thought--

The common prefix has to be the prefix of *all* the lines. So maybe start with the first line and find its whitespace prefix. Then iterate over all the other lines. For each line, you compute the number of characters the two have in common and you find the minimum N of that. Then you keep the first N characters of the prefix.

In pseudocode...

```
fn count_characters_in_common(s1: &str, s2: &str) -> usize {
    s1.chars().zip(s2.chars()).take_while(|(c1, c2)| c1 == c2).count()
}

let non_empty_line_iter = s.lines().filter(|&line| !line.trim().is_empty());
let first_line = non_empty_line_iter.next();
let mut prefix = whitespace_prefix_of(first_line);
let shared_prefix_chars = non_empty_line_iter.map(|s| count_characters_in_common(prefix, s)).min();
prefix = &prefix[..shared_prefix_chars];
strip_prefix_from_original_string(prefix);
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: components/dada-validate/src/validate/validator.rs (hunk 1 above)
Comment (XFFXFF) [reply components/dada-validate/src/validate/validator.rs:552]: This will make the code simpler and more efficient, thanks! I'll refactor it latter.

Comment (nikomatsakis) [reply components/dada-validate/src/validate/validator.rs:552]: One note is that the code I wrote counts characters, but for slicing, you really want to count utf-8 width. You can do that in various ways -- e.g. [`len_utf8`](https://doc.rust-lang.org/std/primitive.char.html#method.len_utf8) or the iterator (forgot the name) that gives you both characters and byte offsets.

Comment (XFFXFF) [reply components/dada-validate/src/validate/validator.rs]: I modified the code to count characters to count bytes, I think this should be equivalent to `len_utf8`.

File: components/dada-validate/src/validate/validator.rs [hunk 2]
```diff
@@ -540,32 +540,47 @@ impl<'me> Validator<'me> {
...
+    if let Some(first_line) = non_empty_line_iter.next() {
+        let prefix = first_line
+            .chars()
+            .into_iter()
+            .take_while(|c| c.is_whitespace())
+            .collect::<String>();
+        let common_indent = non_empty_line_iter
+            .map(|s| count_bytes_in_common(prefix.as_bytes(), s.as_bytes()))
+            .min()
+            .unwrap_or(0);
```
Comment (nikomatsakis) [inline components/dada-validate/src/validate/validator.rs:566]: Probably fine, yes. There might be some weird edge case in unicode I don't know about but I guess we can take our chances. =)

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**